hazelcast/src/main/java/com/hazelcast/sql/impl/expression/datetime/Formatter.java

```diff
@@ -0,0 +1,791 @@
+        private final int exponent;
+
+        enum Form { Normal, Exponential, Roman }
+
+        interface Anchorable {
+            boolean isAnchored();
+        }
+
+        enum Pattern implements Anchorable {
+            BR, B, SG, S, MI, M, PL, P, CR, C, TH, th, EEEE, RN;
+
+            static final List<Pattern> SIGN = Arrays.asList(BR, B, SG, S, MI, M, PL, P);
+            boolean isSign() {
+                return SIGN.contains(this);
+            }
+            char getSign(boolean pre, boolean negative) {
+                if (this == BR || this == B) {
+                    return negative ? (pre ? '<' : '>') : ' ';
+                } else if (this == SG || this == S) {
+                    return negative ? '-' : '+';
+                } else if (this == MI || this == M) {
+                    return negative ? '-' : ' ';
+                } else if (this == PL || this == P) {
+                    return negative ? ' ' : '+';
+                }
+                throw new IllegalArgumentException();
+            }
+
+            static final List<Pattern> ANCHORED = Arrays.asList(B, S, M, P, C, TH, th, EEEE, RN);
+            @Override
+            public boolean isAnchored() {
+                return ANCHORED.contains(this);
+            }
+        }
+
+        static class Literal implements Anchorable {
+            final String contents;
+            final boolean anchored;
+
+            /** In the forms: {@code .+}, {@code F.*}, {@code "[^"]*"}, {@code F"[^"]*"}. */
+            Literal(String literal) {
+                anchored = !literal.startsWith("F");
+                int skip = anchored ? 0 : 1;
+                int quote = literal.charAt(skip) == '"' && literal.length() >= skip + 2
+                        && literal.endsWith("\"") ? 1 : 0;
+                contents = literal.substring(skip + quote, literal.length() - quote);
+            }
+
+            @Override
+            public boolean isAnchored() {
+                return anchored;
+            }
+
+            @Override
+            public String toString() {
+                return (anchored ? "" : "F") + '"' + contents + '"';
+            }
+        }
+
+        /**
+         * Stores {@link Literal}, {@link Pattern}, {@link Character} and {@link Integer} to
+         * represent arbitrary text, sign/currency/ordinals/exponent/roman numerals, grouping/
+         * decimal separators, and digit groups respectively.
+         */
+        class Mask extends ArrayList<Object> {
+            final boolean pre;
+            final StringBuilder digitMask = new StringBuilder();
+            final List<Boolean> fillModes = new ArrayList<>();
+            boolean fillMode = true;
+            int offerSign = -1, digits, minDigits;
+            boolean sign;
+            Pattern bracket;
+
+            Mask(boolean pre) {
+                this.pre = pre;
+            }
+
+            void toggleFillMode() {
+                fillMode = !fillMode;
+            }
```

Comment:
FM isn't a toggle, but a flag. This test should pass: 
```java
    @Test
    public void multipleFm() {
        Formatter f = new Formatter("990");
        check(0, f, "   0");
        f = new Formatter("FM990");
        check(0, f, "0");
        f = new Formatter("FMFM990");
        check(0, f, "0");
        f = new Formatter("FMFMFM990");
        check(0, f, "0");
    }
```

Reply:
See the fill mode's definition in the class Javadoc.